src/app/core/device.ts

```diff
@@ -0,0 +1,1036 @@
+                        length: 85000
+                    },
+                    json.name
+                );
+            }
+        } else {
+            console.warn("Could not find device params, using some default values for device size");
+            newDevice = new Device(
+                {
+                    width: 135000,
+                    length: 85000
+                },
+                json.name
+            );
+        }
+        //TODO: Use this to dynamically create enough layers to scroll through
+        //newDevice.__loadLayersFromInterchangeV1(json.layers);
+        //TODO: Use these two generate a rat's nest
+        newDevice.__loadComponentsFromInterchangeV1(json.components);
+        newDevice.__loadConnectionsFromInterchangeV1(json.connections);
+
+        let valve_map, valve_type_map;
+        //Import ValveMap
+        if (json.params.hasOwnProperty("valveMap") && json.params.hasOwnProperty("valveTypeMap")){
+            valve_map = IOUtils.jsonToMap(json.params.valveMap);
+            console.log("Imported valvemap", valve_map);
+            
+
+            console.log("Loaded valvetypemap", json.params.valveTypeMap);
+            valve_type_map = IOUtils.jsonToMap(json.params.valveTypeMap);
+
+
+            console.log(json.params.valveTypeMap, valve_type_map);
+            let valveis3dmap = new Map();
+            for(let [key, value] of valve_type_map){
+                console.log("Setting type:",key, value);
+                switch(value){
+                    case 'NORMALLY_OPEN':
+                        valveis3dmap.set(key, false);
+                        break;
+                    case 'NORMALLY_CLOSED':
+                        valveis3dmap.set(key, true);
+                        break;
+                }
+            }
+            console.log("Imported valvemap", valve_map, valveis3dmap);
+            newDevice.setValveMap(valve_map, valveis3dmap);
+        }
+
+        //TODO: Use this to render the device features
+
+        //Check if JSON has features else mark
+        if (json.hasOwnProperty("features")) {
+            newDevice.__loadFeatureLayersFromInterchangeV1(json.features);
+        } else {
+            //We need to add a default layer
+            let newlayer = new Layer(null, "flow");
+            newDevice.addLayer(newlayer);
+            newlayer = new Layer(null, "control");
+            newDevice.addLayer(newlayer);
+        }
+
+        //Updating cross-references
+        let features = newDevice.getAllFeaturesFromDevice();
+        let feature;
+        for (let i in features) {
+            //console.log("Feature:", features[i]);
+            feature = features[i];
+            if (feature.referenceID != null) {
+                newDevice.updateObjectReference(feature.referenceID, feature.getID());
+            }
+        }
+
+        return newDevice;
+    }
+
+
+    render2D() {
+        return this.__renderLayers2D();
+    }
```

Comment:
This function appears always to result in an error.
Device.render2D -> Device._renderLayers2D -> Layer.render2D -> Layer._renderFeatures2D -> Features.render2D -> only throws an error

Reply:
When is it throwing the error ? when you're loading the device ? That would make a bunch of difference

Reply:
It appears that these functions are never called, so I've deleted them all